src/charm.py

```diff
@@ -79,29 +126,73 @@ def _on_start(self, event: ops.charm.StartEvent) -> None:
                 return
 
         try:
+            logger.debug("opening tcp port")
             self._open_port_tcp(self._port)
         except subprocess.CalledProcessError:
-            self.unit.status = BlockedStatus("failed to open TCP port for MongoDB")
+            self.unit.status = BlockedStatus(
+                "failed to open TCP port for MongoDB")
             return
 
-        if not self._mongo.is_ready():
+        # check if this unit's deployment of MongoDB is ready
+        if not self._mongo.is_ready(as_stand_alone=True):
+            logger.debug("mongoDB not ready, deferring on start event")
             self.unit.status = WaitingStatus("waiting for MongoDB to start")
             event.defer()
             return
 
-        # initialize replica set
-        self.unit.status = MaintenanceStatus("initialising MongoDB replicaset")
-        logger.debug("initialising replica set")
+        # mongod is now active
+        self.unit.status = ActiveStatus()
+
+        # only leader should initialise the replica set
+        if not self.unit.is_leader():
+            return
+
+        # do not initialise replica set until all peers have joined
+        if not self._number_of_expected_units == self._number_of_current_units:
```

Comment:
I may be missing something but am surprised to see two places in the charm where there is a need to check expected/current units and "initialise" the replica set. Since AFAIU the MonogDB replica set needs to be initialized only once and so it could be done just once for example in the start event handler. There after there certainly will be the need to "reconfigure" the replica set every time a unit joins or departs. 

Also it may be simpler to compare the number of units in the replica set and the number of units currently available rather than the number of units currently available and the number of units expected.